Implement Connectivity.java example using Java bindings

vdhyasani17 opened the pull request:
[WIP] This adds a Java version of the connectivity example using Open MPI's Java bindings. It demonstrates basic point-to-point communication and serves as a reference for Java-based MPI applications. The functionality will be the same as connectivity_c.c.

vdhyasani17 commented:
Hi @jsquyres, can you review my PR when you have the chance? Thank you!

jsquyres commented on docs/man-openmpi/man3/MPI_T_pvar_get_info.3.rst:
```diff
@@ -166,7 +166,9 @@ returned in the *bind* parameter may be one of the following:
-For more information see MPI-3 section 14.3.2.
+* ``MPI_T_BIND_MPI_SESSION``: MPI session
+
+For more information see MPI-4 section 14.3.2.
```
The MPI tools chapter moved in MPI-4.

```suggestion
For more information see MPI-4.1 section 15.3.2.
```

jsquyres commented on examples/Connectivity.java:
```diff
@@ -0,0 +1,66 @@
+/*
+ * Test the connectivity between all processes
+ * WIP
```
Remove WIP.

jsquyres commented on examples/Connectivity.java:
```diff
@@ -0,0 +1,66 @@
+        MPI.Init(args);
+
+        /*
+         * MPI_COMM_WORLD is the communicator provided when MPI is
```
```suggestion
         * MPI.COMM_WORLD is the communicator provided when MPI is
```

jsquyres commented on examples/Connectivity.java:
```diff
@@ -0,0 +1,66 @@
+        int myRank = MPI.COMM_WORLD.getRank();
+        int numProcesses = MPI.COMM_WORLD.getSize();
+        Status status;
+        int verbose = 0;
```
Should this be a boolean instead of an integer?

vdhyasani17 replied:
Yes. Not sure why the C example uses an integer. Is that standard C convention?

jsquyres replied:
I think the C example is so old that it's from before MPI supported MPI_C_BOOL.  Would you happen to be inspired enough to change the C example in this PR as well?  (Perhaps in a separate commit than the Java stuff)

vdhyasani17 replied:
Yes, I would be interested after this Java example merges. Thank you!

jsquyres commented on examples/Connectivity.java:
```diff
@@ -0,0 +1,66 @@
+        String processorName = MPI.getProcessorName();
+
+        for (int i = 0; i < numProcesses; i++) {
+            if (myRank == i) { /* Find current process */
```
I don't know the Java ecosystem at all -- is it more common for single-line comments to be `//` instead of `/* ... */`?

jsquyres commented on examples/Connectivity.java:
```diff
@@ -0,0 +1,66 @@
+                        System.out.printf("Checking connection between rank %d on %s and rank %d\n", i, processorName,
+                                j);
+
+                    int[] dataBuffer = { myRank }; /*
```
There's mixing of styles here: end-of-line comments and before-the-line comments.  We should probably be consistent.

In the majority of the rest of the Open MPI code base (and in `examples/connectivity.c`), we use before-the-line comments.  That might be a good choice here, for consistency.

vdhyasani17 replied:
Ok, I will fix that to strictly use before-the-line comments.

jsquyres commented on examples/Connectivity.java:
```diff
@@ -0,0 +1,66 @@
+                                                    * Data buffer to be sent to next process (in this case, only send
+                                                    * rank num)
+                                                    */
+                    int[] recBuffer = new int[1]; /* Data buffer that current process receives from previous process */
```
In `connectivity_c.c`, the variables are named `rank` and `peer`.  If this example is meant to be the Java version of the C example, it might be good to make them similar in spirit -- e.g., have the variable names and meanings agree.

vdhyasani17 replied:
Got it. I will rename them to `rank` and `peer`.

jsquyres commented on examples/Connectivity.java:
```diff
@@ -0,0 +1,66 @@
+
+                    /* Process must send message first then wait to receive to avoid deadlock */
+                    MPI.COMM_WORLD.send(dataBuffer, 1, MPI.INT, j, myRank);
+                    MPI.COMM_WORLD.recv(recBuffer, 1, MPI.INT, j, j);
```
Heh.  I guess we hadn't really looked closely at `connectivity_c.c` before, because technically this is a deadlock -- using a blocking send to send to yourself *might* block and *might not*.  In this case, it (most likely) won't because the message is so short -- but it definitely *could*.

This example -- and the C example -- should likely be fixed.

vdhyasani17 replied:
Ok, interesting. Should I use non-blocking `iSend` and `iRecv` instead to fix this?

jsquyres replied:
It might be simpler to use MPI_Sendrecv() -- that is guaranteed not to deadlock (e.g., even if the corresponding MPI_Send() followed by MPI_Recv() could have deadlocked).

jsquyres commented on examples/Connectivity.java:
```diff
@@ -0,0 +1,66 @@
+                }
+            } else if (myRank > i) {
+                int[] dataBuffer = { myRank };
+                int[] recBuffer = new int[1];
```
Is there a reason to use a fixed array for `dataBuffer` and an alloc'ed buffer for `recBuffer`?  I.e., why the difference?

vdhyasani17 replied:
I used a fixed array for data buffer because I thought it would be fine prepopulating the array with `myRank` since I knew I would be sending that, but I will keep it consistent in this next push. Also, I will be using `IntBuffer` instead of `int[]` for iSend and iRecv methods.

jsquyres reviewed (changes requested):
I'm not quite sure how the extra mpi.h.in / python changes came into this PR.  Can you rebase to the current tip of `main` and avoid including Howard's commit here?

vdhyasani17 commented:
Sorry, I had some issues with rebasing, so all of my commits are now underneath the comments.

jsquyres commented:
> Sorry, I had some issues with rebasing, so all of my commits are now underneath the comments.

Rebasing is no problem -- I slice and dice my commits via interactive rebasing all the time.  😄

jsquyres commented:
Go ahead and resolve the github conversations that you have addressed -- makes it easier to see what has been done.

Also, can you squash your Java commits down to a single commit?  When we merge this PR, we probably only need to see a single commit in the final git history showing the addition of this example.  Thanks!

vdhyasani17 commented on examples/Connectivity.java:
```diff
@@ -0,0 +1,86 @@
+                     */
+                    Request sendReq = MPI.COMM_WORLD.iSend(rank, 1, MPI.INT, j, myRank);
+                    Request recvReq = MPI.COMM_WORLD.iRecv(peer, 1, MPI.INT, j, j);
+                    sendReq.waitFor();
+                    recvReq.waitFor();
```
Ok, do you recommend that I use nonblocking `iSend` and `iRecv` with `waitAll`, or that I use the blocking (I believe it is blocking at least) `sendRecv`?

jsquyres replied:
Let's go simple as use MPI_SENDRECV.

vdhyasani17 replied:
Got it.

vdhyasani17 replied:
How exactly does `sendRecv` work under the hood to not cause deadlocks? Do the `send` and `recv` calls for certain processes get reordered dynamically by MPI?

jsquyres replied:
Crassly, it essentially turns into an IRECV + ISEND + WAITALL.  But you could also imagine a more optimized implementation.

jsquyres replied:
I stand corrected -- the OMPI implementation is essentially IRECV + SEND + WAIT.  See https://github.com/open-mpi/ompi/blob/main/ompi/mpi/c/sendrecv.c.in.

jsquyres commented:
You might want to edit your commit message -- it's currently showing as 3 commit messages in one.

vdhyasani17 commented:
Ok, by showing three commit messages, do you mean that it's showing 'signoff: Vasista Dhyasani' three times?

jsquyres commented:
> It also shows that Luke was the author..? I don't think you want that.
> 
> There's also unrelated changes in the commit. Did you end up squashing Luke's commits into yours?

In the commit message, yes.

But it also shows that Luke was the author..? I don't think you want that.

There's also unrelated changes in the commit. Did you end up squashing Luke's commits into yours?

vdhyasani17 commented:
Ok, so I should just have one commit message (not a combination of commits) that summarizes what I've been doing, and change the author to me?

vdhyasani17 commented:
I apologize for the few day delay. I believe I now rebased correctly so that there is only one commit message, and I am the author of this PR. I also added this example to examples/Makefile I believe.

jsquyres reviewed (approved):
I pushed a drive-by bug fix to this PR (an unquoted shell variable use in configure, regarding the configuration of the Java bindings on macOS).

Thanks for the contribution!

1. Can you cherry-pick these 2 commits to the v5.0.x branch?
2. Do the docs at https://docs.open-mpi.org/en/v5.0.x/features/java.html look sufficient to you?  I ask because we don't get a lot of Java users; you're obviously one, so I figured I'd ask. 😄 
3. I noticed one minor annoyance when I was testing your fix: our `mpirun` recognizes the `java` token (per https://docs.open-mpi.org/en/v5.0.x/features/java.html) to set the classpath, etc.  But it only recognizes the naked token `java` -- it doesn't recognize it if it's a relative or absolute path.  It might be a nice minor enhancement to notice if the basename of the argv[0] of the executable to be launched is `java` -- not just the naked token `java`.  If you're so inclined, that would be a great feature for another PR.
